helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -255,11 +261,14 @@ private void refreshBaseline(ResourceControllerDataProvider clusterData,
       Map<HelixConstants.ChangeType, Set<String>> clusterChanges, Map<String, Resource> resourceMap,
       final CurrentStateOutput currentStateOutput) throws HelixRebalanceException {
     LOG.info("Start calculating the new best possible assignment.");
+    LatencyMetric partialRebalanceMetric = new PartialRebalanceMetric("PartialRebalanceLatency");
```

Comment:
Regardless of the code style and details, my concerns are:
Say you have implemented every single metric very well and they all can be visualized and analyzed, but there's no way to link separate metrics together from this design. In more detail, suppose our goal is to understand out of the entire pipeline, what's the latency breakdown for each major process?
Without the linkability, it's incomplete and biased since you cannot just add average latency metric up and get the total latency. 
If your goal is to understand the average latency of each step, regardless of how the total latency is, this design is fine. My concern is we cannot get much out of it,

For example, the waged rebalance pipeline consists of 3 steps, (step0, step1, step2). In one of the running process, the total latency is 100s (it breaks down into 5s, 80s, and 15s each); In the other running process, the total latency is 1000s (breaks down into 900s, 50s, 50s). In such an extreme case, how could we use the step0's standlone latency for the explanations?

Reply:
@i3wangyi 
I'm not sure if I understand your concern entirely.

It actually is meant to be like that - this is just one metric we are adding. Here I've only added one metric just as an example, but every "step" of the pipeline like the way you referred to, will have an implementation. 

So you could say that we will at least have a Metric object for each step. We will also have a metric for the entire step. We'll emit appropriate gauges/counters/Histogram for more detailed analysis.

In short, we want to do exactly what you're describing. But at the code level, we won't have any dependency between metrics or hierarchy since JMX will treat every metric equally anyways (whether it represents step 0 or the entire duration of the pipeline, they are just another metric to JMX). Hope that clears things up.